src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -496,22 +501,47 @@ public void afterAnalysis(
       CommandEnvironment env,
       BuildRequest request,
       BuildOptions buildOptions,
-      Iterable<ConfiguredTarget> configuredTargets) {
+      AnalysisResult analysisResult) {
     if (remoteOutputsMode != null && remoteOutputsMode.downloadToplevelOutputsOnly()) {
       Preconditions.checkState(actionContextProvider != null, "actionContextProvider was null");
       boolean isTestCommand = env.getCommandName().equals("test");
       TopLevelArtifactContext artifactContext = request.getTopLevelArtifactContext();
-      ImmutableSet.Builder<ActionInput> filesToDownload = ImmutableSet.builder();
-      for (ConfiguredTarget configuredTarget : configuredTargets) {
+      Set<ActionInput> filesToDownload = new HashSet<>();
+      for (ConfiguredTarget configuredTarget : analysisResult.getTargetsToBuild()) {
         if (isTestCommand && isTestRule(configuredTarget)) {
-          // When running a test download the test.log and test.xml.
+          // When running a test download the test.log and test.xml. These are never symlinks.
           filesToDownload.addAll(getTestOutputs(configuredTarget));
         } else {
-          filesToDownload.addAll(getArtifactsToBuild(configuredTarget, artifactContext).toList());
-          filesToDownload.addAll(getRunfiles(configuredTarget));
+          fetchSymlinkDependenciesRecursively(
+              analysisResult.getActionGraph(),
+              filesToDownload,
+              getArtifactsToBuild(configuredTarget, artifactContext).toList());
+          fetchSymlinkDependenciesRecursively(
+              analysisResult.getActionGraph(),
+              filesToDownload,
+              getRunfiles(configuredTarget));
         }
       }
-      actionContextProvider.setFilesToDownload(filesToDownload.build());
+      actionContextProvider.setFilesToDownload(ImmutableSet.copyOf(filesToDownload));
+    }
+  }
+
+  private static void fetchSymlinkDependenciesRecursively(
+      ActionGraph actionGraph, Set<ActionInput> builder, List<Artifact> inputs) {
+    for (Artifact input : inputs) {
+      // Only fetch recursively if we don't have the file to avoid visiting nodes multiple times.
+      if (builder.add(input)) {
+        fetchSymlinkDependenciesRecursively(actionGraph, builder, input);
+      }
+    }
+  }
+
+  private static void fetchSymlinkDependenciesRecursively(
+      ActionGraph actionGraph, Set<ActionInput> builder, Artifact artifact) {
+    ActionExecutionMetadata action = (ActionExecutionMetadata) actionGraph.getGeneratingAction(artifact);
```

Comment:
what happens if a top-level artifact is a tree artifact generated by a template action? I think that can happen?

If so, then it should be ok to ignore non-ActionExecutionMetadata instances, since templates shouldn't insensitively propagate their inputs (although I guess theoretically there could be a template action that just symlinked its inputs?). Could pull #mayInsensitivelyPropagateInputs() up to ActionAnalysisMetadata too.

Reply:
Good catch. Right now that leads to a Bazel crash. Fixed!

The main issue right now are SymlinkAction and SymlinkTreeAction, but in general, any remotely run action can return a symlink in the open-source remote execution API. Therefore, this will eventually need to be rewritten to follow symlinks at execution time. The current design mixes up action execution and output fetching, and separating them should actually result in a nicer implementation, because we can then remove the code that injects remote execution flags into Skyframe.

So yeah, this is a short-term fix to get cc_binary to work properly, and there's more work needed.

Reply:
Would you mind adding a TODO for that somewhere appropriate?

Reply:
Sorry, now that I understand this better, could you add a test exhibiting this behavior? I assume it would be a Starlark action that created a symlink on the remote worker?

The behavior of creating symlinks on the remote worker seems very problematic to me. What happens if the action creates an intermediate file and then its final output is a symlink to that file? We'll just download a dangling symlink? Is there any chance of getting rid of this symlinking feature altogether?